[FLINK-1159] Case style anonymous functions not supported by Scala API

stefanobaghino opened the pull request:
The proposed API extension methods would allow developers to pass a pattern matching anonymous function so that they are applied on a `DataSet` or `DataStream`; many methods defined on the `DataSet` and `DataStream` APIs don't support those functions due to the overloading[[1]][[2]]; pattern matching anonymous functions allow a very idiomatic approach in Scala to *decompose tuples, case classes and collections*.

The PR does not pollute the original `DataSet` and `DataStream` APIs but is provided as an optional set of extensions methods, implemented via implicit conversions and made available to the developer by explicitly importing the required package, e.g.:

```scala
import org.apache.flink.api.scala.extensions.acceptPartialFunctions
env.fromElements('a -> 1, 'b -> 2).mapWith {
  case (key, value) =>
    ... // key and value are now available and have a sensible name
}
```

[1]: https://groups.google.com/d/msg/scala-user/3oHnDEl1UsM/dDNir9BsiG4J
[2]: http://www.scala-lang.org/files/archive/spec/2.11/06-expressions.html#overloading-resolution

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/acceptPartialFunctions/package.scala:
```diff
@@ -0,0 +1,174 @@
+  *         case Point(x, y) => (x, y)
+  *       }.flatMapWith {
+  *         case (x, y) => Seq('x' -> x, 'y' -> y)
+  *       }.groupingBy {
```
Isn't it called `keyingBy`?

tillrohrmann replied:
Sorry forget my comment.

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/acceptPartialFunctions/package.scala:
```diff
@@ -0,0 +1,174 @@
+  * e.g.
+  * {{{
+  *   object Main {
+  *     import org.apache.flink.api.scala.extensions._
```
This won't work. You have to import `org.apache.flink.api.scala.extensions.acceptPartialFunctions._`

stefanobaghino replied:
Yup, leftover from the previous implementation, thanks for pointing it out, I'll fix this.

tillrohrmann commented on flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/extensions/acceptPartialFunctions/package.scala:
```diff
@@ -0,0 +1,133 @@
+  *         case Point(x, y) => (x, y)
+  *       }.flatMapWith {
+  *         case (x, y) => Seq('x' -> x, 'y' -> y)
+  *       }.groupingBy {
```
Shouldn't this be called `keyingBy`?

stefanobaghino replied:
Copy/paste error, good catch! I'll fix this.

tillrohrmann commented on flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/extensions/acceptPartialFunctions/package.scala:
```diff
@@ -0,0 +1,133 @@
+  * e.g.
+  * {{{
+  *   object Main {
+  *     import org.apache.flink.api.scala.extensions._
```
import shouldn't work if I'm not mistaken

tillrohrmann commented on flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/extensions/acceptPartialFunctions/package.scala:
```diff
@@ -0,0 +1,133 @@
+
+  }
+
+  implicit class OnKeyedStream[T: TypeInformation, K](stream: KeyedStream[T, K]) {
```
What is with the `fold` operation?

tillrohrmann commented on flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/extensions/acceptPartialFunctions/package.scala:
```diff
@@ -0,0 +1,133 @@
+  * }}}
+  *
+  */
+package object acceptPartialFunctions {
```
I think the `ConnectedStreams` and `WindowedStreams` are missing.

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/acceptPartialFunctions/package.scala:
```diff
@@ -0,0 +1,174 @@
+          fun(left.to[Seq], right.to[Seq])
+      }
+
+  }
```
`GroupedDataSet` is missing

tillrohrmann replied:
`CrossDataSet` is missing

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/acceptPartialFunctions/package.scala:
```diff
@@ -0,0 +1,174 @@
+      */
+    def groupingBy[K: TypeInformation: ClassTag](fun: T => K): GroupedDataSet[T] =
+      ds.groupBy(fun)
+
```
`combineGroupWith` is not supported.

tillrohrmann commented:
Thanks for your contribution @stefanobaghino. I really like this feature a lot :-)

Currently, the implementation is not complete, because the supported set of API calls is not complete. 

With the current packaging structure one would always have to import `org.apache.flink.api.scala.extensions.acceptPartialFunctions._`. I would rather like to import the following to get partial function support `org.apache.flink.api.scala.extensions.acceptPartialFunctions` or if I want to import all extensions: `org.apache.flink.api.scala.extensions._`. We could achieve this by introducing an `extensions` package object which does something like:

```
package object extensions {
  implicit def acceptPartialFunctions[T: TypeInformation](ds: DataStream[T]):
    DataStreamWithPartialFunctionSupport[T] = {
    new DataStreamWithPartialFunctionSupport[T](ds)
  }
```

What do you think?

tillrohrmann commented:
It would also be great to add some tests to make sure that the import is working. Furthermore, it would be great to add documentation for the extension feature.

stefanobaghino commented:
Thanks @tillrohrmann, I'll fix the errors in the comments and add the missing methods and extensions.

Regarding the import mode, I agree with you. I started off with just one `DataSet` but then had to support many, I'll try to go back to the original design while retaining multiple ´DataSet´ subtype extensions.

The only thing I'm not very convinced of from your snippet is about returning the extended `DataStreamWithPartialFunctionSupport` class. Implicit conversions have no runtime cost, so maybe it would be better to just return the `DataSet` to make the code more compact and readable. What do you think?

Regarding the tests: I actually have some tests in place locally but basically they're just copies of the original tests on the operators. Should I commit them as well?

Regarding the docs: absolutely, I wasn't sure about this. Where do you think it would be better to put them? Should I add a new chapter under the programming guides?

tillrohrmann commented:
How do you want to do the implicit conversion if you return a `DataSet[T]`. If it makes the code more readable, then I think it's a good idea :-)

We should have a test, which makes sure that the implicit conversion works when you import the corresponding package. That should be enough.

I think we could add the documentation to the streaming guide. We could add a new page with the methods and link it from the scala tab of [DataStream Transformations](https://ci.apache.org/projects/flink/flink-docs-master/apis/streaming/index.html#datastream-transformations) in the streaming guide.

stefanobaghino commented:
Just the streaming guide? Not on both the streaming and batch?

tillrohrmann commented:
Of course, you're right. Also in the batch guide :-)

StephanEwen commented:
Looks very nice in my opinion.

Could you remove the `ClassTag` context bounds? We recently removed them from `DataStream`, because they are not needed.

stefanobaghino commented:
:+1: @StephanEwen I started working on the extension before the removal, will fix this as well, thanks for the feedback.

stefanobaghino commented:
@tillrohrmann I got mixed up reading your proposal and just got what you meant with the `extensions.acceptPartialFunctions` implicit conversion, thanks for the tip, I'll add it to the PR as well.

stefanobaghino commented:
This should cover the missing implementations. I also included the tests I used to test the functionality, let me know if you prefer a wider coverage. I'll provide the docs ASAP.

stefanobaghino commented:
@StephanEwen I had to restore some of the context bounds on `ClassTag` to make it compile, apparently the delegated methods use them; I've rebased with the latest changes on the master before putting them back in place.

StephanEwen commented:
@stefanobaghino Ah, yes, for `DataSet` you need the `ClassTag`, for `DataStream` they should not be needed...

tillrohrmann commented:
There are scalastyle violations in the code:

```
error file=/home/travis/build/apache/flink/flink-tests/src/test/scala/org/apache/flink/api/scala/extensions/AcceptPFFilterITCase.scala message=File must end with newline character
error file=/home/travis/build/apache/flink/flink-tests/src/test/scala/org/apache/flink/api/scala/extensions/AcceptPFFlatMapITCase.scala message=File must end with newline character
error file=/home/travis/build/apache/flink/flink-tests/src/test/scala/org/apache/flink/api/scala/extensions/AcceptPFMapITCase.scala message=File must end with newline character
```

stefanobaghino commented:
@tillrohrmann Yes, I've already fixed them locally, thanks for the notice and for pinging me. I'm working on the documentation, sorry if it's taking me a little bit longer than expected but I had little time this week. I want to complete the PR by Sunday evening.

stefanobaghino commented:
I've squashed the commits, I don't know what I was thinking about. :disappointed: 
This should more or less be it, I'd just like to add more tests to cover all the operators.

stefanobaghino commented:
@tillrohrmann The tests failure seem to be flaky, I've re-run them all on our fork and they're all green now (after a couple of retries).

StephanEwen commented:
This pull request adds a lot of tests (which is actually good), but all tests fire up a cluster to execute many programs. But need to somehow get this down, as these "fire up mini cluster" tests have made our build times explode.

Is actual program execution needed here? Or is it sufficient to see that the partial function for a certain transformation creates such a transformation?

stefanobaghino commented:
Thanks for the feedback. Basically the tests so far have been a carbon copy of the ones run on the operators I delegate to from the extensions. I agree that this approach is a little bit heavy and would like to make it lighter but perhaps I can use some help on this.

I've made a few changes that perhaps go in the right direction in my latest commit ([source here](https://github.com/radicalbit/flink/blob/3e6c58c84c977299d05ef9c9675ecda31368731a/flink-tests/src/test/scala/org/apache/flink/api/scala/extensions/AcceptPFMapITCase.scala)).

Here's a snippet:

```scala
class AcceptPFMapITCase extends TestLogger {

  private val optimizer = new Optimizer(new Configuration)
  private val unusedResultPath = "UNUSED"

  def isTransformation(n: PlanNode): Boolean =
    !n.isInstanceOf[SourcePlanNode] && !n.isInstanceOf[SinkPlanNode]

  private def getOptimizerTransformationNodes(env: ExecutionEnvironment): Iterable[OptimizerNode] =
    for {
      node <- optimizer.compile(env.createProgramPlan("UNUSED")).getAllNodes
      transformation = node.getOptimizerNode if isTransformation(node)
    } yield transformation

  @Test
  def testIdentityMapperWithBasicType(): Unit = {
    val env = ExecutionEnvironment.getExecutionEnvironment
    val ds = CollectionDataSets.getStringDataSet(env)
    val identityMapDs = ds.mapWith(identity)
    identityMapDs.writeAsText(unusedResultPath, WriteMode.OVERWRITE)
    val nodes = getOptimizerTransformationNodes(env)
    assertTrue("The plan should contain 1 transformation", nodes.size == 1)
    assertTrue("The transformation should be a map", nodes.forall(_.isInstanceOf[MapNode]))
  }

  /* [...] */
```

If this approach sounds fine, I can apply it to the other tests as well, along with a wider coverage of the extension methods.

Apart from the approach in general, I have a couple of doubts:
* is it safe to share an `Optimizer` among several tests (i.e. no state involved)?
* is it enough to test the transformation or should I test the input and output types as well?

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/acceptPartialFunctions/OnDataSet.scala:
```diff
@@ -0,0 +1,104 @@
+  def mapPartitionWith[R: TypeInformation: ClassTag](fun: Seq[T] => R): DataSet[R] =
+    ds.mapPartition {
+      (it, out) =>
+        out.collect(fun(it.to[Seq]))
```
Does `it.to[Seq]` materializes the `iterator`? If so, then this is not so good because you can run out of memory.

stefanobaghino replied:
[Yes, it does.](http://www.scala-lang.org/api/current/index.html#scala.collection.Traversable@to[Col[_]]:Col[A]) I thought in this context I was fairly safe from OOMs but I'll refactor to make it work on each item of the collection individually. Thanks!

tillrohrmann replied:
I think that `it.to[Seq]` can be problematic, since it can happen that an underlying `Vector` is returned here. This means that the whole iterator will be materialized. It is better imo to define `fun: Iterator[T] => R`.

tillrohrmann replied:
@stefanobaghino, the idea of `mapPartition` is to process all elements but also to have the possibility to access them all from within one mapPartition call. Therefore, it should get an `Iterator[T]` as input instead of iterating over the iterator and calling the mapPartition for each element.

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/acceptPartialFunctions/OnDataSet.scala:
```diff
@@ -0,0 +1,104 @@
+  def reduceGroupWith[R: TypeInformation: ClassTag](fun: Seq[T] => R): DataSet[R] =
+    ds.reduceGroup {
+      (it, out) =>
+        out.collect(fun(it.to[Seq]))
```
Same question here with the materialization of the iterator.

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/acceptPartialFunctions/OnGroupedDataSet.scala:
```diff
@@ -0,0 +1,75 @@
+  def reduceGroupWith[R: TypeInformation: ClassTag](fun: Seq[T] => R): DataSet[R] =
+    ds.reduceGroup {
+      (it, out) =>
+        out.collect(fun(it.to[Seq]))
```
Materialization?

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/acceptPartialFunctions/OnGroupedDataSet.scala:
```diff
@@ -0,0 +1,75 @@
+  def combineGroupWith[R: TypeInformation: ClassTag](fun: Seq[T] => R): DataSet[R] =
+    ds.combineGroup {
+      (it, out) =>
+        out.collect(fun(it.to[Seq]))
```
Materialization?

tillrohrmann commented on docs/apis/scala_api_extensions.md:
```diff
@@ -0,0 +1,392 @@
+To use this extension exclusively, you can add the following `import`:
+
+{% highlight scala %}
+import org.apache.flink.api.scala.extensions.acceptPartialFunctions
```
Does this really work? Don't you have to import `o.a.f.api.scala.extensions.acceptPartialFunctionsOnDataSet` etc.?

stefanobaghino replied:
Yes, I wrote the docs before testing and rewriting the method signatures; good catch, thanks. I'll try to find a way to make a single import for all `acceptPartialFunctions` methods (see my reply to the next comment).

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/package.scala:
```diff
@@ -0,0 +1,201 @@
+    * }}}
+    *
+    */
+  implicit def acceptPartialFunctionsOnDataSet[T: TypeInformation](ds: DataSet[T]): OnDataSet[T] =
```
I'm wondering whether we shouldn't overload the method `acceptPartialFunctions` with the different data set types instead of having a different method name for the different data sets (grouped, normal, etc.). I don't think that the user will want to enable partial function support on such a fine grained scale. Either he wants partial function support or not.

stefanobaghino replied:
I've tried but the implicit resolution doesn't seem to be smart enough to recognize which overloaded implicit definition to pick, so I had to fall back to this. I can make a couple of attempts using a common object with several methods or perhaps using type classes.

tillrohrmann replied:
Ah ok, I didn't know that Scala would have problems with implicits and method overloading. If it is not possible or it gets too complicated, then simply leave it as it is.

tillrohrmann replied:
Can you remember what the problem was? I quickly changed the method names and on a first glance the tests still seem to pass.

stefanobaghino replied:
Ok, in the end the problem was a clash between the `acceptPartialFunctions` package and implicit conversions. I've solved it by moving the actual implementations under the `impl` package under `acceptPartialFunctions`; now the whole set of conversions can be imported with the `acceptPartialFunctions` name. I was wrong regarding the issue on the resolution of overloaded implicit conversions.

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/package.scala:
```diff
@@ -0,0 +1,201 @@
+    *   }
+    * }}}
+    *
+    */
```
JavaDocs does not fit to method since it only works on `JoinDataSet`.

tillrohrmann commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/package.scala:
```diff
@@ -0,0 +1,201 @@
+    *   }
+    * }}}
+    *
+    */
```
Same here with the JavaDocs.

tillrohrmann commented on flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/extensions/acceptPartialFunctions/OnWindowedStream.scala:
```diff
@@ -0,0 +1,78 @@
+  def applyWith[R: TypeInformation](initialValue: R)
+                                   (foldFunction: (R, T) => R,
+                                    windowFunction: (K, W, R) => TraversableOnce[R]):
+      DataStream[R] =
```
Formatting seems to be a bit off here.

tillrohrmann commented on flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/extensions/acceptPartialFunctions/OnWindowedStream.scala:
```diff
@@ -0,0 +1,78 @@
+    */
+  def applyWith[R: TypeInformation](initialValue: R)
+                                   (foldFunction: (R, T) => R,
+                                    windowFunction: (K, W, R) => TraversableOnce[R]):
```
Why does the `windowFunction` work on a single `R` element and not on all elements of a window?

stefanobaghino replied:
The implementation iterates over each `R` item. However, this approach seems to contrast with [the remarks you have on `it.to[Seq]`](https://github.com/apache/flink/pull/1704#discussion_r55998933); I will restore the `Iterator` here as well. Thanks!

tillrohrmann commented on flink-streaming-scala/src/main/scala/org/apache/flink/streaming/api/scala/extensions/package.scala:
```diff
@@ -0,0 +1,202 @@
+  implicit def acceptPartialFunctionsOnWindowedStream[T, K, W <: Window](
+      ds: WindowedStream[T, K, W]) =
+    new OnWindowedStream[T, K, W](ds)
+
```
The same comments as for the extensions for the `DataSet` apply here.

tillrohrmann commented:
@stefanobaghino, really good work :-) I think we're close to get this merged. I had some minor comments.

Concerning the testing, I agree with @StephanEwen that it's not really necessary to execute a complete program for each new extension method. Instead, I think it is sufficient to check that the right operator/stream transformation has been instantiated. The operators and stream transformations should be already well tested. For the `DataSet` methods you could do something like

```
val identityMapDs = ds.mapWith(identity)
assertTrue(identityMapDs.javaSet.isInstanceOf[MapOperator[String, String]])
```

And for the `DataStream` methods

```
val identity = stream.mapWith{case x => x}
assertTrue(identity.javaStream.getTransformation.asInstanceOf[OneInputTransformation[Int, Int]]
      .getOperator.isInstanceOf[StreamMap[Int, Int]])
```

stefanobaghino commented:
@tillrohrmann Thanks! Outstanding review, it's great to have some guidance when approaching a new project; thank you for the tips on testing as well, it turned out to be way more simple then I thought. :) I'll make sure to apply the changes you mentioned, review the documentation and rewrite and expand the tests.

stefanobaghino commented:
@tillrohrmann I just have one minor concern regarding the tests: would `isInstanceOf[StreamMap[Int, Int]]` work? Wouldn't the generic type parameters (the two `Int`s) be erased at runtime?

tillrohrmann commented:
Yes I think they will be removed at runtime. Thus, I guess it should also be fine to test for `isInstanceOf[StreamMap[_, _]]`. If you also want to check the input/output types properly, then you can use the `TypeInformations` which are accessible from the `OneInputTransformation`.

stefanobaghino commented:
A quick status update: unfortunately I have had no time to work on this in the past week, I plan to get back to the fixes on Thursday or Friday.
In the meanwhile, I have a small doubt on the usage of `Iterator`: they are indeed very good to allow the user to have access both to the whole `DataSet`/`DataStream` or accessing it one item at a time; however, they are not particularly useful when used with the case-style partial functions: they offer an edge to de-structure a single item like a tuple or a collection like `Seq` (e.g. using the `_ +: rest` operator to only get the item after the first).
Are we sure we want to keep the `Iterator`? Is there an advantage in having an `Iterator` with this extension? I see to possible solutions: 

1. the easy one: having two methods, one materializing the `Iterator` into a collection and another one accessing the items one a time: the only issue with this would be the need to have to methods with distinct names (otherwise we would be back to square one); this means the user can use the case-style functions to destructure the collection or each item separately; otherwise we can
2. adopt a slightly more sophisticated solution: wrap the `Iterator` in a `Stream`, which is lazy but also fully destructurable in case-style functions (e.g.: using the `#::` operator). This would require some work as the `Iterator` is stateful with regards of the traversal while the `Stream` is not and we can't just use a naive solution or the semantic difference could lead to some nasty bugs in user code.

tillrohrmann commented:
Your 2. proposal looks more natural to me because what you receive as input to a user function is in fact a stream of data. So if this solution is feasible, then I would vote for it.

stefanobaghino commented:
@tillrohrmann I moved forward on the batch extensions: I added support for anonymous partial functions on `where` and `equalsTo` for joins and co-group operations. I also deleted the old tests and provided a full set of unit tests for each method under the same source root (`flink-scala`).

Now I have to work on the streaming and adjust the Scaladoc and documentation. I hope next week I'll have something ready to close this PR. Let me know if you prefer to have the work done so far pushed on this PR already.

tillrohrmann commented:
Sounds great @stefanobaghino. I think you can push your work to this PR as well since it is all related to the partial function support. Looking forward having partial function support :-)

stefanobaghino commented:
Note on the tests of the streaming extensions: I couldn't find a more specific class than `SingleOutputStreamOperator[_]`, thus assertion may not be very meaningful. However, I'd leave them in place to make sure further work on the extension won't break them at compile time.

StephanEwen commented on flink-scala/src/main/scala/org/apache/flink/api/scala/extensions/impl/acceptPartialFunctions/OnDataSet.scala:
```diff
@@ -0,0 +1,111 @@
+  * @param ds The wrapped data set
+  * @tparam T The type of the data set items, for which the type information must be known
+  */
+class OnDataSet[T: TypeInformation](ds: DataSet[T]) {
```
I think you do not need the `TypeInformation` context bound here. Functions should never need the implicit type info for the input (that is contained in the DataSet already), only ever for the return type.

StephanEwen commented:
Looks pretty good. I noticed three remaining things:

  - Sometimes, the package and the directory trees are different (example: `OnCoGroupDataSetTest.scala`). While Scala allows that, we usually keep them in sync.

  - The classes that implement the "withX" functions sometimes have context bounds that I think they should not require (see inline comment above). Would be great to remove them, because removing them later makes it API breaking.

  - We need to decide how to label this for the future: stable, or evolving. I would suggest to use `@PublicEvolving` on all the involved classes for now.

stefanobaghino commented:
@StephanEwen Thanks for the feedback, I've addressed the points you highlighted and took the time to add some missing Scaladoc to the `DataStream` extension.

StephanEwen commented:
Looks good to me. Will merge this finally!

Thanks for the hard work and the many iterations here...

stefanobaghino commented:
It's been a pleasure, thanks for the patience and for guiding me through the iterations. :smiley:
